drivers/gpu/drm/bridge/cdns-mhdp.c

```diff
@@ -778,6 +783,54 @@ static int load_firmware(struct cdns_mhdp_device *mhdp)
 	return 0;
 }
 
+static void mhdp_check_link(struct cdns_mhdp_device *mhdp)
+{
+	struct drm_connector *conn = &mhdp->connector;
+	u8 status[DP_LINK_STATUS_SIZE];
+	bool hpd_state;
+	int hpd_event;
+	int ret;
+
+	/* Nothing to check if there is no link */
+	if (!mhdp->link_up)
+		return;
+
+	hpd_event = cdns_mhdp_read_event(mhdp);
+
+	/* Geting event bits failed, send HPD event */
+	if (hpd_event < 0) {
+		dev_warn(mhdp->dev, "%s: read event failed: %d\n",
+			 __func__, hpd_event);
+		return;
+	}
+
+	hpd_state = !!(hpd_event & DPTX_READ_EVENT_HPD_STATE);
+
+	/* No point the check the link if HPD is down (cable is unplugged) */
+	if (!hpd_state)
+		return;
+
+	/* No display reconfiguration between link check and property setting */
+	mutex_lock(&conn->dev->mode_config.mutex);
+
+	/* Check if the link is still up */
+	ret = drm_dp_dpcd_read_link_status(&mhdp->aux, status);
+
+	/* If dpcd read fails, we must assume the link is down too */
+	if (ret >= 0) {
+		if (drm_dp_channel_eq_ok(status, mhdp->link.num_lanes) &&
+		    drm_dp_clock_recovery_ok(status, mhdp->link.num_lanes))
+			/* Link is Ok, no need to set link status property */
+			goto out;
+	}
+
+	/* Link is broken, so indicate it with the link status property */
+	drm_connector_set_link_status_property(conn, DRM_MODE_LINK_STATUS_BAD);
```

Comment:
Other than the cosmetic comments above, looks good. And you're able to test this code path too? I guess you are with the analyzer.